### PR TITLE
Dialect and Rules api

### LINKED ISSUE #805: Expose supported dialects for import
As a developer using SQLFluff in other projects, it would be helpful to have a list of supported dialects available for import.

Currently I maintain a list of dialects inferred from the project source:

```python
supported_dialects = [
    "bigquery", "mysql", "teradata", "postgres", "snowflake", "exasol", "exasol_fs"
]
```

It would be great to replace this with:

```python
from sqlfluff import SUPPORTED_DIALECTS
```

### PR DESCRIPTION
**Pull request by alanmcruickshank**
This resolves #805 .

I've upgraded both the dialects and rules readouts to use NamedTuples for consistency. This also adds tests and examples.

### CODE REVIEW

File: test/api/info_test.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+"""Test using sqlfluff to extract elements of queries."""
+
+import sqlfluff
+
+
+def test__api__info_dialects():
+    """Basic linting of dialects."""
+    dialects = sqlfluff.dialects()
+    assert isinstance(dialects, list)
+    assert ("ansi", "ansi", "nothing") in dialects
```
Comment (NiallRees) [inline test/api/info_test.py:10]: nothing? Should we filter this out as it's really only an internal thing and everything inherits from ansi?

Comment (alanmcruickshank) [reply test/api/info_test.py:10]: what do you mean "filter it out"? You mean not include the inheritance of a dialect?

Comment (NiallRees) [reply test/api/info_test.py:10]: I mean just prevent `"nothing"` from being returned by `sqlfluff.dialects()`. Definitely include the inheritance

Comment (NiallRees) [reply test/api/info_test.py:10]: Or we could give it a meaningful name if it has to stay in? It's just a bit confusing atm

Comment (alanmcruickshank) [reply test/api/info_test.py:10]: ah  "nothing" isn't being included as a dialect here. `("ansi", "ansi", "nothing")` is a shorthand for ` DialectTuple(label="ansi", name="ansi", inherits_from="nothing")`. Does that make more sense?

Comment (NiallRees) [reply test/api/info_test.py:10]: It does indeed!

**Review by NiallRees — APPROVED** [review]

**Pull request merged**